File: Python3/0338. Counting Bits.py
class Solution(object):
    def countBits(self, num):
        """
        :type num: int
        :rtype: List[int]
        """
#         res = [0]*(num+1)

#         for i in range(1,num+1):
#             res[i] = (res[i-1]+1) if i%2 else res[i//2]
        
#         return res

        # res = [0]*(num+1)

        # mask = 1
        # while(mask<=num):
        #     for i in range(1,num+1):
        #         res[i] += (mask&i)!=0
        #     mask <<= 1

        # return res
        
        arr = [0]
        bit = 1
        while((1<<bit)<num):
            bit += 1
        for i in range(1,bit+1):
            for j in range(1<<(i-1)):
                arr.append(1+arr[j])
        return arr[:num+1] if len(arr)>num else arr+[1]

File: C++/0338. Counting Bits.cpp
class Solution {
public:
    vector<int> countBits(int num) {
        vector<int> arr={0};
        int bit = 1;
        while((1<<bit)<num){
            bit += 1;
        }
        for(int i=1;i<=bit;++i){
            for(int j=0;j<(1<<(i-1));++j){
                arr.push_back(1+arr[j]);
            }
        }
        if(arr.size()>num){
            arr.assign(arr.begin(),arr.begin()+num+1);
        }
        else{
            arr.push_back(1);
        }
        return arr;
    }
};
